VM Migrator to move VM's between GCP zones, regions & projects

suchitpuri opened the pull request:
* GCP VM Migrator lets you move your VMs between different regions, VPCs, Zone and Projects within GCP.  This use case was not supported by M4CE or other tools we had to build it from scratch for some of our customers.  

* Some of the features supported are  
  - Cloning the entire subnet and migrating VM’s from subnet in one region to another. 
  - Support for retaining the primary and alias ips of a VM.
  - Upgrading the machine types on the fly to optimize cost.
  - Ensuring metadata, labels, tags, network interfaces, disks, service accounts and VM properties are retained during migration.
  - Support to migrate machines running on sole tenant node groups to a different node group.

* Eng Council Bug Id: 167568528

iht commented on tools/vm-migrator/Makefile:
```diff
@@ -0,0 +1,66 @@
+# Copyright 2020 Google LLC. This software is provided as-is,
```
Copyright 2021?

iht reviewed (commented):
I am missing some more structure for the code of this tool; the Python code is all in the same directory as the documentation, Makefile, etc.

I would move the Python code to a subdirectory (or more than one), to create one or more Python modules (e.g. with a __init__.py file), the Python modules could be included in a package with setuptools. With the package installed, the Makefile could just use the corresponding module after installing it, or from the local directory.

iht commented on tools/vm-migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+import machine_type_mapping
+import machine_image
+from ratemate import RateLimit
+rate_limit = RateLimit(max_count=2000, per=100)
```
Blank line between code and imports. Upper case for `rate_limit`?

iht commented on tools/vm-migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+    """
+    Create Instance method create a new GCP VM from the machine image
+    """
+
```
Remove blank line (no blank lines after docstring). Also, add a dot at the end of the docstring.

https://www.python.org/dev/peps/pep-0257/#one-line-docstrings

iht commented on tools/vm-migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+    # Reserve the static ip before creating the instance
+    # If we get ip variable set, we expect to use the same ip in the
+    # destination subnet
+    if (ip):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+        config['networkInterfaces'][0]['networkIP'] = get_ip(
+            compute, project, name, get_region_from_zone(zone))
+
+    if (node_group and get_updated_node_group(node_group)):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+        config['scheduling'] = get_updated_node_group(node_group)['scheduling']
+
+    i = 1
+    if ((len(alias_ip_ranges) > 0)):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+
+def wait_for_instance(compute, project, zone, name):
+    """
+    Function to wait for creation of machine image
```
Dot at the end of one-line docstring.

iht commented on tools/vm-migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+           machine_type=None,
+           wait=True):
+    """
+    Main functin to create the instance
```
Misspelled "function".

Dot at the end of one-line docstring.

iht commented on tools/vm-migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+            (instance_name, instance_name))
+        return instance_name
+
+    except Exception as ex:
```
`Exception` here is too broad. I would rather use a custom exception and `sys.exit` rather than re-raising the exception (and would let any other exception unhandled).

iht commented on tools/vm-migrator/machine_image.py:
```diff
@@ -0,0 +1,96 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file is used to create a machine image for an instance
```
Dot at the end of one-line docstring.

iht commented on tools/vm-migrator/machine_image.py:
```diff
@@ -0,0 +1,96 @@
+import time
+import googleapiclient.discovery
+import logging
+from ratemate import RateLimit
```
Blank line after imports.

Upper case for `rate_limit`?

iht commented on tools/vm-migrator/machine_image.py:
```diff
@@ -0,0 +1,96 @@
+from ratemate import RateLimit
+rate_limit = RateLimit(max_count=2000, per=100)
+
+# machineImage
```
Two blank lines around functions.

iht commented on tools/vm-migrator/machine_image.py:
```diff
@@ -0,0 +1,96 @@
+        return None
+
+
+# [START wait_for_operation]
```
Unnecessary comment?

iht commented on tools/vm-migrator/machine_image.py:
```diff
@@ -0,0 +1,96 @@
+        time.sleep(30)
+
+
+# [END wait_for_operation]
```
Unnecessary comment?

iht commented on tools/vm-migrator/machine_image.py:
```diff
@@ -0,0 +1,96 @@
+            wait_for_operation(compute, project, name)
+        logging.info('Machine Image %s Created' % (name))
+        return name
+    except Exception as exc:
```
`Exception` too broad, please see other comments for same code pattern in other files.

iht commented on tools/vm-migrator/machine_type_mapping.py:
```diff
@@ -0,0 +1,89 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file is used to maintain mapping of different node groups
```
Dot at the end of docstring.

iht commented on tools/vm-migrator/machine_type_mapping.py:
```diff
@@ -0,0 +1,89 @@
+This file is used to maintain mapping of different node groups
+"""
+
+find = {
```
UPPER CASE? Maybe more descriptive name? What is this dictionary exactly?

iht commented on tools/vm-migrator/node_group_mapping.py:
```diff
@@ -0,0 +1,22 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file is used to maintain mapping of different node groups
```
Dot at the end of docstring.

iht commented on tools/vm-migrator/node_group_mapping.py:
```diff
@@ -0,0 +1,22 @@
+This file is used to maintain mapping of different node groups
+"""
+
+find = {
```
Upper case? More descriptive name?

iht commented on tools/vm-migrator/run.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
`/usr/bin/env bash`

Bash might not be necessarily installed in `/bin`

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file deals with operations on subnets
```
Dot at the end of one-line docstring.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+"""
+This file deals with operations on subnets
+"""
+import googleapiclient.discovery
```
Please reorder imports following the order given at https://www.python.org/dev/peps/pep-0008/#imports

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+    compute = get_compute()
+    subnet = "https://www.googleapis.com/compute/beta/" + subnet
+
+    if (ip.endswith("/32")):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+                                   region=region,
+                                   filter='(subnetwork="' + subnet +
+                                   '") (address="' + ip + '")').execute()
+    if (ips.get("items") and len(ips.get("items")) == 1):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+    result = compute.instances().list(project=project,
+                                      zone=zone,
+                                      maxResults=10000).execute()
+    if (not result.get('items')):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+    if (not result.get('items')):
+        result = {'items': []}
+
+    if (zone_2):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+        result_zone_2 = compute.instances().list(project=project,
+                                                 zone=zone_2,
+                                                 maxResults=10000).execute()
+    if (zone_3):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+
+    mydict = []
+
+    if (result_zone_2.get("items") and zone_2):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+    if (result_zone_2.get("items") and zone_2):
+        result['items'] = result['items'] + result_zone_2.get("items")
+
+    if (result_zone_3.get("items") and zone_3):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+    for instances in result['items']:
+
+        headers = fields.headers
+        if ((instances['networkInterfaces'][0]['subnetwork']).endswith(subnet)):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+                            disks['deviceName']))
+
+            aliasIps = instances['networkInterfaces'][0].get('aliasIpRanges')
+            if (aliasIps):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+                        csv['range_name_' +
+                            str(i + 1)] = aliasIps[i]['subnetworkRangeName']
+
+            if (instance.is_hosted_on_sole_tenant(instances)):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+
+        logging.info("Successfully written %i records to %s" %
+                     (len(mydict), file_name))
+        print("Successfully written %i records to %s" %
```
Having print and a logger in the same place is strange. Users can choose the logging output level in the stdout if they need to.

iht commented on tools/vm-migrator/subnet.py:
```diff
@@ -0,0 +1,243 @@
+                                   region=region,
+                                   filter='subnetwork="' + subnet +
+                                   '"').execute()
+    if (ips.get("items")):
```
Remove unnecessary parentheses.

iht commented on tools/vm-migrator/zone_mapping.py:
```diff
@@ -0,0 +1,23 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file is used to maintain mapping of different node groups
```
Dot at the end of one-line docstring.

iht commented on tools/vm-migrator/zone_mapping.py:
```diff
@@ -0,0 +1,23 @@
+This file is used to maintain mapping of different node groups
+"""
+
+find = {
```
UPPER CASE? Maybe more descriptive name for this dictionary?

suchitpuri commented:
@iht can you please review the code structure now, i have tried to fix all the comments except the exception part as i need to verify how will it behave in multi threaded environment.

iht commented:
@suchitpuri Starting now to review the changes. I will reply back once I am done with that review.

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+from . import fields
+from csv import DictReader
+from csv import DictWriter
+
```
Two blank lines around function defitions

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                                 (machine_image_name))
+                    logging.info("Machine image %i out of %i completed" %
+                                 (tracker, count))
+                except Exception as exc:
+                    logging.error(
+                        'machine image creation generated an exception: %s' %
+                        (exc))
+
```
TBD: handle custom exceptions

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                    logging.info('%r machine deleted sucessfully' %
+                                 (instance_name))
+                    logging.info("%i out of %i deleted" % (tracker, count))
+                except Exception as exc:
+                    logging.error(
+                        'machine deletion generated an exception: %s' % (exc))
```
TBD: handle custom exceptions

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+            for row in csv_dict_reader:
+                parsed_link = instance.parse_self_link(row['self_link'])
+                for i in range(9):
+                    if (row['disk_name_' + str(i + 1)] != ''):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                                 (disk_name))
+                    logging.info("%i out of %i deleted" % (tracker, count))
+                except Exception as exc:
+                    logging.error(
+                        'disk deletion generated an exception: %s' % (exc))
+
```
TBD: handle custom exceptions

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                    logging.info('%r machine shutdown sucessfully' %
+                                 (machine_name))
+                    logging.info("%i out of %i shutdown" % (tracker, count))
+                except Exception as exc:
+                    logging.error(
+                        'machine shutdown generated an exception: %s' % (exc))
```
TBD: handle custom exceptions

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                    logging.info('%r machine started sucessfully' %
+                                 (machine_name))
+                    logging.info("%i out of %i started up" % (tracker, count))
+                except Exception as exc:
+                    logging.error(
+                        'machine strating generated an exception: %s' % (exc))
```
TBD: handle custom exceptions

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                # This support upto 4 ip ranges but they can be easily extended
+                for i in range(4):
+                    alias_range = {}
+                    if (row['range_name_' + str(i + 1)] != ''):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                    if (row['range_name_' + str(i + 1)] != ''):
+                        alias_range['subnetworkRangeName'] = row['range_name_' +
+                                                                 str(i + 1)]
+                    if (row['alias_ip_name_' + str(i + 1)]):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                    if (row['alias_ip_name_' + str(i + 1)]):
+                        alias_range['aliasIpName'] = row['alias_ip_name_' +
+                                                         str(i + 1)]
+                    if (row['alias_ip_' + str(i + 1)]):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                # This supports up to 4 disks
+                disk_names = {}
+                for i in range(9):
+                    if (row['device_name_' + str(i + 1)] != ''):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                                                                      str(i + 1)]
+
+                node_group = None
+                if (row['node_group'] and row['node_group'] != ''):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                    logging.info('%r machine created sucessfully' %
+                                 (machine_name))
+                    logging.info("%i out of %i created " % (tracker, count))
+                except Exception as exc:
+                    logging.error(
+                        'machine creation generated an exception: %s' % (exc))
```
TBD: handle custom exceptions

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        (destination_file, len(filtered)),
+        default='no')
+
+    if (overrtie_file):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+            if (row['name'] in machine_names_to_filter):
+                filtered.append(row)
+
+    overrtie_file = query_yes_no(
```
overwrite_file? (possible variable name misspelling)

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+            # Find the reserved alias ips
+            for i in range(4):
+                ip_name = row.get('alias_ip_name_' + str(i + 1))
+                if (ip_name != '' and row['range_name_' + str(i + 1)] == ""):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+         file_name='export.csv',
+         filter_file_name='filter.csv'):
+    """
+    The main method to trigger the VM migration
```
Dot at the end of single-line docstring

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+                        level=numeric_level)
+
+    logging.info("executing step %s" % (step))
+    if (step == 'prepare_inventory'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        logging.info("Preparing the inventory to be exported")
+        subnet.export_instances(project, source_zone, source_zone_2,
+                                source_zone_3, source_subnet, "source.csv")
+    if (step == 'filter_inventory'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        logging.info("filtering out the inventory")
+        overrite_file = filter_records("source.csv", filter_file_name,
+                                       file_name)
+        if (overrite_file):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        subnet.export_instances(project, source_zone, source_zone_2,
+                                source_zone_3, source_subnet, "source.csv")
+        logging.info("filtering out the inventory")
+        overrite_file = filter_records("source.csv", filter_file_name,
```
overwrite_file? (possible misspelling of variable name)

suchitpuri replied:
fixed

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        else:
+            logging.info("File %s was not overriten" % (file_name))
+
+    if (step == 'shutdown_instances'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+            "Are you sure you want to shutdown the (%s) instances present in the inventory ?"
+            % count,
+            default='no')
+        if (shutdown_response):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+            logging.info("Shutting down the instances")
+            bulk_instance_shutdown(file_name)
+
+    if (step == 'start_instances'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        start_response = query_yes_no(
+            "Are you sure you want to start the instances present in the inventory ?",
+            default='no')
+        if (start_response):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+            logging.info("Starting the instances")
+            bulk_instance_start(file_name)
+
+    if (step == 'create_machine_images'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        bulk_image_create(project, source_zone, source_subnet,
+                          machine_image_region, file_name)
+
+    if (step == 'delete_instances'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+            "Are you sure you want to delete the (%s) instances present in the inventory ?"
+            % count,
+            default='no')
+        if (response):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        else:
+            logging.info("Not deleting any instances")
+
+    if (step == 'clone_subnet'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        subnet.duplicate(project, subnet_name, source_region, target_region)
+        logging.info("Subnet sucessfully cloned in the provided region")
+
+    if (step == 'create_instances'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        bulk_create_instances(file_name, target_subnet, True)
+        logging.info("Instances created successfully")
+
+    if (step == 'create_instances_without_ip'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        bulk_create_instances(file_name, target_subnet, False)
+        logging.info("Instances created successfully")
+
+    if (step == 'release_ip_for_subnet'):
```
Remove unnecessary parentheses

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,502 @@
+        subnet.release_ip(project, source_region, source_subnet)
+        logging.info("ips present in %s released subcessfully" % (file_name))
+
+    if (step == 'release_ip'):
```
Remove unnecessary parentheses

iht reviewed (commented):
+1 to reorganizing the Python code as a Python module.

Please move `requirements.txt` and `pylintrc` to the same directory as `setup.py`, those are not part of the Python module.

I have found some files that seem to be configuration for VS Code. I think we should not distribute that kind of project/config files, so users can easily import the code with any other IDE/editor.

About using custom exceptions, I am not sure what are the implications for a multi-threaded application. In principle, whether you handle a custom exception or all exceptions and then raise other exceptions, the behaviour should be the same.

Please see the inline comments for suggestions for changes.

suchitpuri commented on tools/vm-migrator/.vscode/.ropeproject/config.py:
```diff
@@ -0,0 +1,114 @@
+# The default ``config.py``
```
makes sense, i checked in it by mistake, removing it

suchitpuri replied:
moved `requirements.txt` and `pylintrc` to the same directory as setup.py

suchitpuri commented on tools/vm-migrator/migrator.log:
```diff
@@ -0,0 +1,227 @@
+2021-01-06 20:47:27,629  INFO executing step prepare_inventory
```
Removing log file, it will be automatically generator for the users

suchitpuri commented on tools/vm-migrator/migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+from . import machine_image
+from ratemate import RateLimit
+
+rate_limit = RateLimit(max_count=2000, per=100)
```
done

suchitpuri commented on tools/vm-migrator/migrator/disk.py:
```diff
@@ -0,0 +1,72 @@
+                    delete_operation['name'])
+            return disk_name
+        else:
+            raise Exception(
+                "Can't delete the disk as machine image not found")
+    except Exception as ex:
+        logging.error(ex)
+        print(ex)
+        raise ex
```
correct, but because this is called from within a `concurrent.futures.ThreadPoolExecutor` if one of the instances raise an error it will kill the entire program, ideally we would want to log it mark the executor as failed and . move ahead with other execturs

suchitpuri replied:
using custom exception here for better clarity

suchitpuri commented on tools/vm-migrator/migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+                                           instance=instance_name).execute()
+        wait_for_zonal_operation(compute, project, zone, result['name'])
+        return instance_name
+    except Exception as ex:
+        logging.error(ex)
+        print(ex)
+        raise ex
+
```
removed the print statement, but if i remove handling `Exception` i am not sure which custom exception to raise since the [documentation](https://googleapis.github.io/google-api-python-client/docs/dyn/compute_beta.instances.html#start) to use `compute.instances().start` does not mention any custom exception it raises hence handling everything and re-raising it to mark the executor as failed.

suchitpuri commented:
removed the unnecessary parenthesis and added custom exceptions wherever possible, since the source python library does not specify the exception i am using `Exception` to catch errors in that and again raise them to make sure the ThreadPoolExecutor framework picks them and marks them as failed.

suchitpuri commented:
@iht please review

iht commented:
@suchitpuri Starting review of new changes now.

mik-laj commented on tools/vm-migrator/Readme.md:
```diff
@@ -0,0 +1,94 @@
+
+| Variable      | Value                              |
+| ------------- | ---------------------------------- |
+| PROJECT          | GCP Project Name                  |
```
```suggestion
| `PROJECT`          | GCP Project Name                  |
```
See:  https://developers.google.com/style/code-in-text

mik-laj commented on tools/vm-migrator/Readme.md:
```diff
@@ -0,0 +1,94 @@
+| PROJECT          | GCP Project Name                  |
+| MACHINE_IMAGE_REGION | The region where you want to store your machine images e.g us-central1 |
+| INPUT_CSV | The filename which will serve as the input to migration process, some of the steps use this filename to export the inventory data  |
+|SOURCE_SUBNET | The source subnet in form projects/\<project-id\>/regions/\<region\>/subnetworks/\<subnet-name\> |
```
```suggestion
|SOURCE_SUBNET | The source subnet in form `projects/\<PROJECT_ID\>/regions/\<rEGiON\>/subnetworks/\<SUBNET_NAME\>` |
```
See: https://developers.google.com/style/placeholders

mik-laj commented on tools/vm-migrator/Readme.md:
````diff
@@ -0,0 +1,94 @@
+
+## Upgrading Machine Types
+
+If you are planing to migrate machines to a different region and in the process of migration you want to optimize the machine types based on the usage you can do that on the fly by providing the source and destination machine type mappings in ```machine_type_mapping.py``` e.g 'n1-standard-1': 'n1-standard-4' would upgrade all the machines running with n1-standard-1 to n1-standard-4
````
```suggestion
If you are planing to migrate machines to a different region and in the process of migration you want to optimize the machine types based on the usage you can do that on the fly by providing the source and destination machine type mappings in ```machine_type_mapping.py``` e.g `n1-standard-1`: `n1-standard-4` would upgrade all the machines running with `n1-standard-1` to `n1-standard-4`
```

mik-laj commented on tools/vm-migrator/Readme.md:
```diff
@@ -0,0 +1,94 @@
+
+| Variable      | Value                              |
+| ------------- | ---------------------------------- |
+| `PROJECT`          | GCP Project Name                  |
```
In other places, it is also worth marking constants as code.  I just corrected one place as an example.
https://developers.google.com/style/code-in-text

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,507 @@
+                    logging.info('%r machine shutdown sucessfully',
+                                 machine_name)
+                    logging.info('%i out of %i shutdown', tracker, count)
+                except (GCPOperationException, Exception) as exc:
```
Same comment about handling only `Exception` here.

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,507 @@
+                    logging.info('%r machine started sucessfully',
+                                 machine_name)
+                    logging.info('%i out of %i started up', tracker, count)
+                except (GCPOperationException, Exception) as exc:
```
Same comment about handling only `Exception` here.

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,507 @@
+                    logging.info('%r machine created sucessfully',
+                                 machine_name)
+                    logging.info('%i out of %i created ', tracker, count)
+                except (GCPOperationException, Exception) as exc:
```
Same comment about handling only `Exception` here.

iht commented on tools/vm-migrator/migrator/instance.py:
```diff
@@ -0,0 +1,401 @@
+        result = shutdown_instance(compute, project, zone, instance_name)
+        wait_for_zonal_operation(compute, project, zone, result['name'])
+        return instance_name
+    except (GCPOperationException, Exception) as ex:
```
Same comment about handling only `Exception` here.

iht commented on tools/vm-migrator/migrator/instance.py:
```diff
@@ -0,0 +1,401 @@
+                                           instance=instance_name).execute()
+        wait_for_zonal_operation(compute, project, zone, result['name'])
+        return instance_name
+    except (GCPOperationException, Exception) as ex:
```
Same comment about handling only `Exception` here.

iht commented on tools/vm-migrator/migrator/instance.py:
```diff
@@ -0,0 +1,401 @@
+        else:
+            raise NotFoundException(
+                'Cant Delete the instance as machine image not found')
+    except (GCPOperationException, NotFoundException, Exception) as ex:
```
Same comment about handling only `Exception` here.

iht commented on tools/vm-migrator/migrator/instance.py:
```diff
@@ -0,0 +1,397 @@
+def get_region_from_zone(zone):
+    match = re.search(r"(\w+)-(\w+)-(\w+)", zone)
+    if (len(match.groups()) != 3):
+        raise Exception('Invalid Zone Format')
```
What kind of exception is this? If you raise a custom exception, it would be possible to handle that type of exception isolated from other types of exceptions. If you raise a generic `Exception`, it cannot be handled isolatedly.

iht commented on tools/vm-migrator/migrator/instance.py:
```diff
@@ -0,0 +1,401 @@
+            instance_name, instance_name)
+        return instance_name
+
+    except (GCPOperationException, Exception) as ex:
```
Same comment about handling only `Exception` here.

iht reviewed (changes requested):
I think we are almost here :).

The only remaining changes that should be done are the following:

- In some places, several exceptions are captured together, but the handling code is generic. As all are of type `Exception`, I think only `Exception` should be captured.
- In some places, generic `Exception` are raised, it is impossible to handle those exceptions isolated from any other one. I would raise a custom type of exception in those places.

Please see the inline comments for the specific locations.

Thanks!

suchitpuri commented:
Fixed all the exception handling suggestions, hopefully this should be it

iht commented on tools/vm-migrator/migrator/subnet_region_migrator.py:
```diff
@@ -0,0 +1,507 @@
+
+import argparse
+import logging
+from migrator.exceptions import GCPOperationException
```
The linter is failing because `GCPOperationException` is never used in this file.

iht reviewed (changes requested):
The build failed because of the linter (one unused import).

iht reviewed (approved):
LGTM.

There is still one pending change to pass the linter (the PR cannot be merged until it passes). Other than that, it looks good. Thanks!

suchitpuri commented:
@iht all checks passed, please merge the branch

iht commented:
Thanks @suchitpuri for the contribution and @mik-laj for the improvements in the documentation.
